common/src/main/java/io/netty/util/concurrent/ReentrantImmediateEventExecutor.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.concurrent;
+
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * Executes {@link Runnable} objects in the caller's thread. If the {@link #execute(Runnable)} is reentrant it will be
+ * queued until the original {@link Runnable} finishes execution.
+ * <p>
+ * All {@link Throwable} objects thrown from {@link #execute(Runnable)} will be swallowed and logged. This is to ensure
+ * that all queued {@link Runnable} objects have the chance to be run.
+ */
+public final class ReentrantImmediateEventExecutor extends ImmediateEventExecutor {
```

Comment:
No longer propagating exceptions sounds fine to me, in general. Since "normal" executors will never propagate it's not too strange for this executor not to propagate. It may be useful to allow plugging in other behavior (as is done with Thread.UncaughtExceptionHandler for Threads). Using the default uncaught exception handler may be reasonable, except that the handlers are normally called just before thread death. If the thread's life continues, it may break some assumption (although I have a _really_ hard time figuring out what).

However, test code may be negatively impacted. I wouldn't be surprised if some test code is doing an assert within a Runnable and letting it propagate up the stack to fail the test. Maybe that is enough reason to have `ImmediateEventExecutor`.